src/Datasource/ConnectionRegistry.php

```diff
@@ -68,24 +69,26 @@ protected function _throwMissingClassError(string $class, ?string $plugin): void
      * If a callable is passed as first argument, The returned value of this
      * function will be the result of the callable.
      *
-     * @param callable|object|string $class The classname or object to make.
+     * @param object|string $class The classname or object to make.
      * @param string $alias The alias of the object.
      * @param array<string, mixed> $config An array of settings to use for the datasource.
-     * @return object A connection with the correct settings.
+     * @return \Cake\Datasource\ConnectionInterface A connection with the correct settings.
      */
-    protected function _create(callable|object|string $class, string $alias, array $config): object
+    protected function _create(object|string $class, string $alias, array $config): ConnectionInterface
     {
-        if (is_callable($class)) {
-            return $class($alias);
-        }
+        if (is_string($class)) {
+            unset($config['className']);
 
-        if (is_object($class)) {
-            return $class;
+            /** @var \Cake\Datasource\ConnectionInterface */
+            return new $class($config);
         }
 
-        unset($config['className']);
+        if ($class instanceof Closure) {
```

Comment:
If we're going to support a hidden Closure object, I think it should be part of the parent `_create()` parameter type hint.

Reply:
Maybe we can implement closure support in the base class and not require every _create() to handle it.

Reply:
> If we're going to support a hidden Closure object, I think it should be part of the parent _create() parameter type hint.

Can't have `Closure|object` as type. It leads to `Fatal error: Type Closure|object|string contains both object and a class type, which is redundant`.

Reply:
I wonder if we can just drop support for `Closure`. `ConnectionRegistry` seems to be the only one supporting it.

I think the idea was to potentially delay creating an expensive connection instance only when needed. Not sure how useful that is in practice.

Reply:
Ugh.. `LogEngine::_create()` also accepts `callable`. In general I am trying to get rid of `callable`s where not necessary.

Reply:
Yea, makes sense that it throws an error. I wasn't 100% sure.

Reply:
> I wonder if we can just drop support for Closure. ConnectionRegistry seems to be the only one supporting it.

I don't think we should be dropping `Closure` support. It can be useful when defining connections/loggers with the DI container, at least until decide on whether or not all those things should be defined in the DI container entirely.

Reply:
> It can be useful when defining connections/loggers with the DI container, at least until decide on whether or not all those things should be defined in the DI container entirely.

Makes sense, I hadn't considered DIC usage. 

This should be good to merge then.